Implement response type converters

rdlaitila opened the pull request:
work for timheuer/alexa-skills-dotnet#128

stoiveyp commented on Alexa.NET/Response/Directive/VoicePlayerSpeakDirective.cs:
```diff
@@ -5,11 +5,14 @@
+    public class VoicePlayerSpeakDirective : IProgressiveResponseDirective
     {
-        public VoicePlayerSpeakDirective(Ssml.Speech speech):this(speech.ToXml())
+        public VoicePlayerSpeakDirective()
```
The other parameterless constructors you've added have been internal - is there a reason this one is public?

rdlaitila replied:
Nope just an oops on my part
```suggestion
        internal VoicePlayerSpeakDirective()
```

stoiveyp commented on Alexa.NET/Response/Response.cs:
```diff
@@ -17,12 +20,12 @@ public class ResponseBody
         public bool? ShouldEndSession { get; set; } = false;
 
-        [JsonProperty("directives", NullValueHandling = NullValueHandling.Ignore)]
+        [JsonProperty("directives", NullValueHandling = NullValueHandling.Ignore, ItemConverterType = typeof(DirectiveConverter))]
```
This should be on the IDirective definition itself, that way we have the advantage of this new deserialization regardless of where it's used or any future changes of ResponseBody

rdlaitila replied:
Agreed I'll move the converter to IDirective instead

rdlaitila commented:
@stoiveyp thanks for the quick review. I wanted to get this pr out soon while my use case issue is fresh in my mind.

I can certainly retool most of those tests to deserialize directly against the example json documents. I agree that is probably the better solution. I'll get that under review in another commit sometime shortly.

One other nagging question for me is the public api for registering new response converters from other consumers. For instance I believe your own skills purchasing nuget needs to register new request types and I believe there are special subscription directives that would want to be handled for response deserialization.

Currently the dead simple way was a public static dictionary containing string -> delegate pairs to obtain a stub object for Newtonsoft.Json to complete the deserialization. If you are happy with that api surface we can go with it otherwise I was toying with the idea of registering new converters perhaps with a similar api as the `RequestConverter.RequestConverters`. Thoughts?

stoiveyp commented:
@rdlaitila I had the same thought about the converters initially, but on reflection the request converters have to be able to handle multiple types per registration because its normally one per interface.

By contrast the converters you've created are much more a one to one mapping. My extensions will have new directives to add, but I don't think a distinct converter will add enough value to justify it - what you've done now seems like the right fit to me.

rdlaitila commented:
@stoiveyp retooling of deserialization tests against example json files is done. Let me know your thoughts, thanks!

stoiveyp commented on Alexa.NET.Tests/ResponseTests.cs:
```diff
@@ -190,6 +192,544 @@ public void RepromptSsmlGeneratesPlainTextOutput()
+
+            Assert.Equal("1.0", deserialized.Version);
+
+            var sessionAttributeSupportedHoriscopePeriods =
```
At this stage we should be able to get supportedHoriscopePeriods as a JObject and then use .Value<bool> for the subsequent assertions.

The reason I say this isn't regarding the dynamic itself - more that using a dynamic allows a certain flexibility in the tests implementation; ideally that's something we want to avoid if people use the tests for information on expected behavior.

rdlaitila replied:
updated to use JObject in cleanup commit

stoiveyp commented on Alexa.NET.Tests/ResponseTests.cs:
```diff
@@ -190,6 +192,544 @@ public void RepromptSsmlGeneratesPlainTextOutput()
             Assert.Equal(speech.ToXml(), ssmlText.Ssml);
         }
 
+        [Fact]
```
Can we please move the test to the bottom - almost the entire file is now part of the PR diff, makes it tricky to ensure there weren't any accidental changes anywhere else.

rdlaitila commented:
@stoiveyp 

reverted Response.cs and Reprompt.cs since those files ended up not having any final changes. Also moved the deserialize response test to bottom of file and moved to use JObject.Value as you suggested.

Let me know if you have any final items. Thanks again!

rdlaitila commented on Alexa.NET/Response/Reprompt.cs:
```diff
@@ -10,15 +11,16 @@ public Reprompt()
 
-        [JsonProperty("outputSpeech", NullValueHandling=NullValueHandling.Ignore)]
+        [JsonProperty("outputSpeech", NullValueHandling = NullValueHandling.Ignore)]
+        [JsonConverter(typeof(OutputSpeechConverter))]
```
reverted to the last commit from master

rdlaitila commented on Alexa.NET/Response/Response.cs:
```diff
@@ -1,5 +1,6 @@
 ﻿using Newtonsoft.Json;
 using System.Collections.Generic;
+using Alexa.NET.Response.Converters;
```
reverted to the last commit from master

german95n commented:
Waiting for this PR! Thank you @rdlaitila  this will help my unit test to check if the SkillResponse is correct or not! ^^
